pyteal/ast/router.py

```diff
@@ -0,0 +1,647 @@
+
+    def __lt__(self, other: "ConflictMapElem"):
+        # compare under same oc condition
+        # can be used to order AST
+        if not isinstance(other, ConflictMapElem):
+            raise TealInputError(
+                "ConflictMapElem can only check conflict with other ConflictMapElem"
+            )
+
+        if self.is_method_call:
+            if not other.is_method_call:
+                return False
+        else:
+            if other.is_method_call:
+                return True
+
+        # either both is_method_call or not
+        # compare on_creation
+        if self.on_creation:
+            return not other.on_creation
+        else:
+            return other.on_creation
+
+    def has_conflict_with(self, other: "ConflictMapElem"):
+        if not isinstance(other, ConflictMapElem):
+            raise TealInputError(
+                "ConflictMapElem can only check conflict with other ConflictMapElem"
+            )
+        if not self.is_method_call and not other.is_method_call:
+            if self.method_name == other.method_name:
+                raise TealInputError(f"re-registering {self.method_name} under same OC")
+            else:
+                raise TealInputError(
+                    f"re-registering {self.method_name} and {other.method_name} under same OC"
+                )
+
+
+ConflictMapElem.__module__ = "pyteal"
+
+
+class ASTConflictResolver:
+    def __init__(self):
+        self.conflict_detect_map: dict[str, list[ConflictMapElem]] = {
+            name: list() for name in dir(OnComplete) if not name.startswith("__")
+        }
+
+    def add_elem_to(self, oc: str, conflict_map_elem: ConflictMapElem):
+        if oc not in self.conflict_detect_map:
+            raise TealInputError(
+                f"{oc} is not one of the element in conflict map, should be one of the OnCompletes"
+            )
+        elems_under_oc: list[ConflictMapElem] = self.conflict_detect_map[oc]
+        for elem in elems_under_oc:
+            if elem.has_conflict_with(conflict_map_elem):
+                raise TealInputError(f"{elem} has conflict with {conflict_map_elem}")
+
+        self.conflict_detect_map[oc].append(conflict_map_elem)
+
+
+ASTConflictResolver.__module__ = "pyteal"
+
+
+class Router:
+    """
+    Class that help constructs:
+    - an ARC-4 app's approval/clear-state programs
+    - and a contract JSON object allowing for easily read and call methods in the contract
+    """
+
+    def __init__(
+        self, name: Optional[str] = None, bare_calls: list[OnCompleteAction] = None
+    ) -> None:
+        """
+        Args:
+            name (optional): the name of the smart contract, used in the JSON object.
+                Default name is `contract`
+        """
+        if bare_calls is None:
+            bare_calls = DEFAULT_BARE_CALLS
+        # TODO need to change API to allow defining bare_call on init
+
+        self.name: str = "Contract" if name is None else name
```

Comment:
@ahangsu A few thoughts here and open to discussion:
* I propose _requiring_ `name` because I believe there is no sensible default.  While nothing prevents re-using the same name across contract definitions, I _think_ we want to encourage naming specific to the use case at-hand.
* https://github.com/algorandfoundation/ARCs/blob/main/ARCs/arc-0004.md#contract-description describes rules for contract naming (quoted below).  Neither the PR nor py-algorand-sdk (`sdk_abi.Contract`) appear to enforce the rules.  Is that by-design because we rely on algod to perform validation or is it unintentional?
> Contract names MUST satisfy the regular expression [_A-Za-z][A-Za-z0-9_]*.

Reply:
good point, I also had some thoughts of requiring name to be required, or Contract name would be duplicated, which is fishy.

I can also add a naming restriction, e.g., regex, to enforce the naming align with arc-004 pattern.

Reply:
Resolving - contract name is now an overridable parameter.